executor/containerdexecutor/executor.go

```diff
@@ -45,33 +48,56 @@ func New(client *containerd.Client, root, cgroup string, networkProviders map[pb
 		networkProviders: networkProviders,
 		cgroupParent:     cgroup,
 		dnsConfig:        dnsConfig,
+		running:          make(map[string]chan error),
+		mu:               &sync.Mutex{},
 	}
 }
 
-func (w containerdExecutor) Exec(ctx context.Context, meta executor.Meta, root cache.Mountable, mounts []executor.Mount, stdin io.ReadCloser, stdout, stderr io.WriteCloser) (err error) {
-	id := identity.NewID()
+func (w containerdExecutor) Run(ctx context.Context, id string, root cache.Mountable, mounts []executor.Mount, process executor.ProcessInfo, started chan<- struct{}) (err error) {
+	if id == "" {
+		id = identity.NewID()
+	}
+
+	startedOnce := sync.Once{}
+	done := make(chan error, 1)
+	w.mu.Lock()
+	w.running[id] = done
+	w.mu.Unlock()
+	defer func() {
+		w.mu.Lock()
+		delete(w.running, id)
+		w.mu.Unlock()
+		close(done)
+		if started != nil {
+			startedOnce.Do(func() {
+				close(started)
+			})
+		}
+	}()
+
+	meta := process.Meta
 
 	resolvConf, err := oci.GetResolvConf(ctx, w.root, nil, w.dnsConfig)
 	if err != nil {
-		return err
+		return sendErr(done, err)
```

Comment:
This is better if wrapped in a single `defer` so there is no risk that wrapping was forgotten.

Reply:
Sure I can do that. I tend to avoid that pattern myself since I have always found the unintentional redefinition of `err` to be problematic and harder to find, ie:
```
func Exec(...) (err error) {
  defer func() {
    done <- err
  }()
  resolvConf, err := oci.GetResolvConf(ctx, w.root, nil, w.dnsConfig)
  if err != nil {
    return err
}
```
would send `nil` on `done` if `GetResolveConf` fails.

Reply:
> would send nil on done if GetResolveConf fails.

No, it does not :)

Reply:
Hah, whoa, TIL.  I tried it on play just to sanity check, I must have misinterpreted the results, so much for sanity :(
I redid the test and of course you are correct, it works fine: https://play.golang.org/p/TvbDMw6q5Z2 I would have sworn I ran into this issue a while ago, maybe older Go versions had this issue?  Dunno, anyway, yeah that will make the change much easier :)

Reply:
maybe this is what you were thinking https://play.golang.org/p/YYOCf97H0il

Reply:
I have no idea, maybe I ran into something like you suggest.  I tried on an old go1.2 runtime, it was fine there also.  So I guess I have just been wrong for years, o'well, glad you set me straight :)